File: Graph/1377.Frog_Position_After_T_Seconds/Python/1377.Frog_Position_After_T_Seconds.py
class Solution(object):
    def frogPosition(self, n, edges, t, target):
        """
        :type n: int
        :type edges: List[List[int]]
        :type t: int
        :type target: int
        :rtype: float
        """
        #============================================#
        # Breath-first search based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Length of edges array #####
        len_edges = len(edges)

        ##### Root-child array #####
        root_child_arry = [[] for _ in range((n + 1))]

        ##### Visit array #####
        visit_arry = [False for _ in range((n + 1))]

        ##### Probability array #####
        prob_arry = [0.0 for _ in range((n + 1))]

        ##### Record queue #####
        record_queue = []


        ###############################################################
        #Breath-first search based loop traversal with recorded arrays
        ##### Step 1: Record indexed-nodes informations with arrays #####
        for edges_idx in range(len_edges):
            (root_child_arry[(edges[edges_idx])[0]]).append((edges[edges_idx])[1]) #Keep updating/recording
            (root_child_arry[(edges[edges_idx])[1]]).append((edges[edges_idx])[0]) #Keep updating/recording

        ##### Step 2: Looped-traversal with recorded arrays #####
        visit_arry[1] = True #Update/Overwrite

        prob_arry[1] = 1.0 #Update/Overwrite

        record_queue.append(1) #Update/Record

        while (record_queue and t):
            len_queue = len(record_queue) #Length of queue

            for queue_idx in range(len_queue):
                curr_node = record_queue.pop(0) #Current node

                edge_cnter = 0 #Edge summary counter

                for next_node in root_child_arry[curr_node]:

                    ##### Check if the next node is visited or not #####
                    if (visit_arry[next_node]):
                        continue

                    edge_cnter += 1 #Keep updating/accumulating

                for next_node in root_child_arry[curr_node]:

                    ##### Check if the next node is visited or not #####
                    if (visit_arry[next_node]):
                        continue

                    visit_arry[next_node] = True #Keep updating/overwriting

                    prob_arry[next_node] = (prob_arry[curr_node] * (1.0 / edge_cnter)) #Keep updating/overwriting

                    record_queue.append(next_node) #Keep updating/recording

                ##### Check if the frog stayed at the same vertex or not #####
                if (edge_cnter):
                    prob_arry[curr_node] = 0.0 #Keep updating/overwriting
                else:
                    pass

            t -= 1 #Keep updating/reducing

        return prob_arry[target]

File: Graph/1377.Frog_Position_After_T_Seconds/C++/1377.Frog_Position_After_T_Seconds.cpp
class Solution {
public:
    double frogPosition(int n, vector<vector<int>> & edges, int t, int target) 
    {
        //=============================================//
        // Input type:                                 //
        //   - int (n-th nodes)                        //
        //   - vector<vector<int>> ref. (edges vector) //
        //   - int (t seconds)                         //
        //   - int (target node value)                 //
        // Return type:                                //
        //   - double (result probability)             //
        //=============================================//

        /*Initialize*/
        ///// Length of edges vector /////
        int len_edges = edges.size();

        ///// Root-child vector /////
        vector<vector<int>> root_child_vect((n + 1));

        ///// Visit vector /////
        vector<bool> visit_vect((n + 1), false);

        ///// Probability vector /////
        vector<double> prob_vect((n + 1), 0.0);

        ///// Record queue /////
        queue<int> record_queue;


        /*Breath-first search based loop traversal with recorded vectors*/
        ///// Step 1: Record indexed-nodes informations with vectors /////
        for (int edges_idx = 0; (edges_idx < len_edges); (edges_idx++)) //Whole
        {
            (root_child_vect[(edges[edges_idx])[0]]).emplace_back((edges[edges_idx])[1]); //Keep updating, recording
            (root_child_vect[(edges[edges_idx])[1]]).emplace_back((edges[edges_idx])[0]); //Keep updating, recording

        } //Whole

        ///// Step 2: Looped-traversal with recorded vectors /////
        visit_vect[1] = true; //Update, Overwrite

        prob_vect[1] = 1.0; //Update, Overwrite

        record_queue.emplace(1); //Update, Record

        while ((! record_queue.empty()) && (t--)) //Cycled-times
        {
            for (int queue_idx = (record_queue.size()); (queue_idx > 0); (queue_idx--)) //Queue-size
            {
                int curr_node = record_queue.front(); record_queue.pop(); //Current node

                int edge_cnter = 0; //Edge summary counter

                for (const int & next_node: root_child_vect[curr_node]) //Next-nodes
                {
                    ///// Check if the next node is visited or not /////
                    if (visit_vect[next_node]) { continue; }

                    (edge_cnter++); //Keep updating, accumulating

                } //Next-nodes

                for (const int & next_node: root_child_vect[curr_node]) //Next-nodes
                {
                    ///// Check if the next node is visited or not /////
                    if (visit_vect[next_node]) { continue; }

                    visit_vect[next_node] = true; //Keep updating, overwriting

                    prob_vect[next_node] = (prob_vect[curr_node] * (1.0 / edge_cnter)); //Keep updating, overwriting

                    record_queue.emplace(next_node); //Keep updating, recording

                } //Next-nodes

                ///// Check if the frog stayed at the same vertex or not /////
                if (edge_cnter) { prob_vect[curr_node] = 0.0; } //Keep updating, overwriting
                else { ; }

            } //Queue-size

        } //Cycled-times

        return prob_vect[target];
    }
};
